File: UzaySimulasyonu(Java)/src/Sınıflar/Arac.java
/**
*
* @author Yunus Emre İnci yunus.inci1@ogr.sakarya.edu.tr
* @since 25.04.2025
* <p>
* Arac Sınıfı;
* Arac nesnelerinin değişkenlerini ve araç ile ilgili tüm fonksiyonları barındırır.
* </p>
*/

package Sınıflar;

public class Arac {
	private String isim;
	private Gezegen cikisGezegeni;
	private Gezegen varisGezegeni;
	private Zaman cikisTarihi;
	private int mesafe;
	
	private String durum;            // Gerekli değişkenler
	private int kalkisaKalanSaat;
	private int hedefeKalanSaat;
	private Zaman hedefeVaracagiTarih;
	private int yolcuSayisi;
	private Gezegen bulunduguGezegen;
	
	// Kurucu fonksiyon
	public Arac(String isim, Gezegen cikisGezegeni, Gezegen varisGezegeni, Zaman cikisTarihi, int mesafe) {
		this.isim = isim;
		this.cikisGezegeni = cikisGezegeni;
		this.varisGezegeni = varisGezegeni;
		this.cikisTarihi = cikisTarihi;
		this.mesafe = mesafe;
		
		durum = "Bekliyor";
		kalkisaKalanSaat = kalkisaKalanSaatHesapla();
		hedefeKalanSaat = hedefeKalanSaatHesapla();
		hedefeVaracagiTarih = hedefeVaracagiTarihHesapla();
		yolcuSayisi = 0;
		bulunduguGezegen = cikisGezegeni;
	}
	
	// Getir fonksiyonları
	public String isimGetir() { return isim; }
	public String durumGetir() { return durum; }
	public String cikisGezegeniIsimGetir() { return cikisGezegeni.isimGetir(); }
	public String varisGezegeniIsimGetir() { return varisGezegeni.isimGetir(); }
	public String hedefeKalanSaatGetir() { 
		if (durum.equals("İmha") ) {
			return "--";
		}
		return String.valueOf(hedefeKalanSaat);
	}
	public String hedefeVaracagiTarihGetir() {
		if (durum.equals("İmha") ) {
			return "--";
		}
		return hedefeVaracagiTarih.toString();
	}
	
	// x saat ilerle fonksiyonu
	public void xSaatIlerle(int x) {
		if (durum.equals("Vardı") || durum.equals("İmha")) return;
		
		hedefeKalanSaat -= x;           
		
		if (durum.equals("Bekliyor")) {
			if (x > kalkisaKalanSaat) {
				x -= kalkisaKalanSaat;
				aracKalti();                // kalkisaKalanSaat sıfırlanırsa araç kalkar
			} else {
				kalkisaKalanSaat -= x;     
				if (kalkisaKalanSaat == 0) {
					aracKalti();
				}
				return;
			}
		}
		if (hedefeKalanSaat <= 0) {         // hedefeKalanSaat sıfırlanırsa araç varır
			aracVardi();
			hedefeKalanSaat = 0;
		}
	}
	
	private void aracKalti() { // Araç kalktı
		durum = "Yolda";       // Durum Yolda
		bulunduguGezegen.nufusGuncelle(-yolcuSayisi); // Bulunduğu Gezegenin nüfusunu güncelle
		bulunduguGezegen = null;
	}
	
	private void aracVardi() { // Araç vardı
		durum = "Vardı";       // Durum Vardı
		bulunduguGezegen = varisGezegeni;
		bulunduguGezegen.nufusGuncelle(+yolcuSayisi); // Bulunduğu Gezegenin nüfusunu güncelle
	}
	
	// yolcuSayisi değişkenini parametreye göre günceller
	public void yolcuSayisiGuncelle(int x) {
		yolcuSayisi += x;
		if (bulunduguGezegen != null) {                   // Bulunduğu Gezegenin nüfusunu güncelle
			bulunduguGezegen.nufusGuncelle(x);
		}
		if (!durum.equals("Vardı") && yolcuSayisi == 0) { // Yolcu Sayısı sıfırlandıysa İmha
			durum = "İmha";
		}
	}
	
	private int kalkisaKalanSaatHesapla() { // Kalkışa kalan saati hesaplayan fonksiyon
		Zaman tarih = cikisGezegeni.tarihGetir();
		int gunSaatMiktari = cikisGezegeni.gunSaatMiktariGetir();
		return tarih.saatFarkiHesapla(gunSaatMiktari, cikisTarihi);
	}
	
	private int hedefeKalanSaatHesapla() { // Hedefe kalan saati hesaplayan fonksiyon
		return mesafe + kalkisaKalanSaat;
	}
	
	private Zaman hedefeVaracagiTarihHesapla() { // Hedefe varacağı tarihi hesaplayan fonksiyon
		Zaman tarih = varisGezegeni.tarihGetir();
		int gunSaatMiktari = varisGezegeni.gunSaatMiktariGetir();
		return tarih.xSaatIlerideTarihKac(gunSaatMiktari, hedefeKalanSaat);
	}
	
	public boolean isimKontrol(String isim) { // Parametre ile isim ayn mı diye kontrol eder
		return this.isim.equals(isim);        // DosyaOkuma için gerekli
	}
}

File: UzaySimulasyonu(C)/src/Arac.c
#include "Arac.h"

//Kurucu fonksiyon
Arac aracOlustur(char* isim, Gezegen cikisGezegeni, Gezegen varisGezegeni, Zaman cikisTarihi, int mesafe) {
    Arac this = (Arac)malloc(sizeof(struct ARAC));
    
    this->isim = strdup(isim);
    this->cikisGezegeni = cikisGezegeni;
    this->varisGezegeni = varisGezegeni;
    this->cikisTarihi = cikisTarihi;
    this->mesafe = mesafe;
    this->durum = strdup("Bekliyor");
    this->yolcuSayisi = 0;
    this->bulunduguGezegen = cikisGezegeni;
    this->kalkisaKalanSaat = kalkisaKalanSaatHesapla(this);
    this->hedefeKalanSaat = this->kalkisaKalanSaat + this->mesafe;
    this->hedefeVaracagiTarih = hedefeVaracagiTarihHesapla(this);

    this->yolcuSayisiGuncelle = &yolcuSayisiGuncelle;
    this->birSaatIlerle = &aracBirSaatIlerle;
    this->yoket = &aracYoket;

    return this;
}

// 1 saat simule eder, Aracın durumlarını kontrol eder.
void aracBirSaatIlerle(Arac this) {
    if (strcmp(this->durum, "Vardi") == 0 || strcmp(this->durum, "Imha") == 0) return;

    this->kalkisaKalanSaat--;
    this->hedefeKalanSaat--;
    
    if (strcmp(this->durum, "Bekliyor") == 0) {
        if (this->kalkisaKalanSaat <= 0) {
            aracKalkti(this);
        }
    } else {
        if (this->hedefeKalanSaat == 0) {
            aracVardi(this);
        }
    }
}

// Arac kalktğında çağırılan fonksiyon
void aracKalkti(const Arac this) {
	free(this->durum);
    this->durum = strdup("Yolda");

	this->bulunduguGezegen->nufus -= this->yolcuSayisi;
	this->bulunduguGezegen = NULL;
}

// Araç vardığında çağırılıan fonksiyon
void aracVardi(const Arac this) {
	free(this->durum);
    this->durum = strdup("Vardi");

	this->bulunduguGezegen = this->varisGezegeni;
	this->bulunduguGezegen->nufus += this->yolcuSayisi;
}

// YolcuSayisini dışarıdan gelen verilere göre düzenleyen fonksiyon, bulunduğu Gezegenin nufusunu da günceller
void yolcuSayisiGuncelle(const Arac this, int x) {
    this->yolcuSayisi += x;
    if (this->bulunduguGezegen != NULL) {
		this->bulunduguGezegen->nufus += x;
	}
    if (strcmp(this->durum, "Vardi") != 0 && this->yolcuSayisi == 0) {
		free(this->durum);
        this->durum = strdup("Imha");
	}
}

// Kalkışa kalan saati hesaplar döndürür
int kalkisaKalanSaatHesapla(const Arac this) {
    return this->cikisTarihi->saatFarkiHesapla(
        this->cikisTarihi,
        this->cikisGezegeni->tarih,
        this->cikisGezegeni->gunSaatMiktari
    );
}

// Hedefe Varacağı tarihi hesaplar oluşturur döndürür
Zaman hedefeVaracagiTarihHesapla(const Arac this) {
    Zaman hedefeVaracagiTarih = this->varisGezegeni->tarih->kopyala(this->varisGezegeni->tarih);
    for (int i = 0; i < this->hedefeKalanSaat; i++) {
        hedefeVaracagiTarih->birSaatIlerle(hedefeVaracagiTarih, this->varisGezegeni->gunSaatMiktari);
    }
    return hedefeVaracagiTarih;
}

// Yıkıcı fonksiyon
void aracYoket(const Arac this) {
    if (this == NULL) return;
    free(this->isim); /// isim kontrol
    free(this->durum);
    this->cikisTarihi->yoket(this->cikisTarihi);
    this->hedefeVaracagiTarih->yoket(this->hedefeVaracagiTarih);
    free(this);
}